octodns/provider/mythicbeasts.py

```diff
@@ -0,0 +1,420 @@
+#
+#
+#
+
+from __future__ import absolute_import, division, print_function, \
+    unicode_literals
+
+import re
+
+from requests import Session
+from logging import getLogger
+
+from ..record import Record
+from .base import BaseProvider
+
+
+def add_trailing_dot(value):
+    '''
+    Add trailing dots to values
+    '''
+    assert value, 'Missing value'
+    assert value[-1] != '.', 'Value already has trailing dot'
+    return value + '.'
+
+
+def remove_trailing_dot(value):
+    '''
+    Remove trailing dots from values
+    '''
+    assert value, 'Missing value'
+    assert value[-1] == '.', 'Value already missing trailing dot'
+    return value[:-1]
+
+
+class MythicBeastsProvider(BaseProvider):
+    '''
+    Mythic Beasts DNS API Provider
+
+    mythicbeasts:
+      class: octodns.provider.mythicbeasts.MythicBeastsProvider
+        zones:
+          my-zone: 'password'
+    '''
+
+    SUPPORTS_GEO = False
+    SUPPORTS_DYNAMIC = False
+    SUPPORTS = set(('A', 'AAAA', 'ALIAS', 'CNAME', 'MX', 'NS',
+                    'SRV', 'SSHFP', 'CAA', 'TXT'))
+    BASE = 'https://dnsapi.mythic-beasts.com/'
+
+    def __init__(self, identifier, passwords, *args, **kwargs):
+        self.log = getLogger('MythicBeastsProvider[{}]'.format(identifier))
+
+        assert isinstance(passwords, dict), 'Missing passwords'
+
+        self.log.debug(
+            '__init__: id=%s, registered zones; %s',
+            identifier,
+            passwords.keys())
+        super(MythicBeastsProvider, self).__init__(identifier, *args, **kwargs)
+
+        self._passwords = passwords
+        sess = Session()
+        self._sess = sess
+
+    def _request(self, method, path, data=None):
+        self.log.debug('_request: method=%s, path=%s data=%s',
+                       method, path, data)
+
+        resp = self._sess.request(method, path, data=data)
+        self.log.debug(
+            '_request:   status=%d data=%s',
+            resp.status_code,
+            resp.text[:20])
+
+        if resp.status_code == 401:
+            raise Exception('Mythic Beasts unauthorized for domain: {}'
+                            .format(data['domain']))
```

Comment:
I generally prefer things to throw a provider specific exception in cases like this.

```python
class MythicBeastsUnauthorizedException(Exception):
    pass

...

    raise MythicBeastsUnauthorizedException(...)
```

Its doubtful anyone will ever try to catch it, but it gives a bit more info when it fires and it usually makes it easier to ensure that the right thing went wrong in tests.